---
title: Analyse transcript abundance in white spruce (*Picea glauca*) mitochondrion
author: Shaun Jackman
output:
  html_document:
    keep_md: yes
    toc: yes
---

# Load libraries
```{r load-libraries}
library(dplyr)
library(ggplot2)
library(knitr) # for kable
library(magrittr)
library(NMF) # for aheatmap
library(readr)
library(uniqtag) # for make_unique

# Bioconductor
library(genomeIntervals)
```

# Functions
```{r functions}
factor_with_default <- function(x, levels) {
	x <- factor(x, levels = names(levels), labels = levels)
	x[is.na(x)] <- levels[names(levels) == ""]
	x
}
factor_switch <- function(x, ...) factor_with_default(x, list(...))
```

```{r knitr}
opts_chunk$set(tidy = FALSE, fig.width = 20, fig.height = 20)
```

# Read the GFF file of annotations
```{r read-gff}
gff_all <- readGff3('pg29mt-scaffolds.gff', isRightOpen = FALSE)
gff <- gff_all[gff_all$type %in% c('gene', 'mRNA', 'rRNA', 'tRNA'),]
ID_Name <- data.frame(
		ID = getGffAttribute(gff, 'ID'),
		Name = getGffAttribute(gff, 'Name'),
		Type = gff$type,
		Size = size(gff),
		stringsAsFactors = FALSE) %>%
	mutate(Name = sub('\\|.*', '', Name)) %>%
	filter(Type == 'gene')
```

# Read the transcript abundance data
```{r read-data}
samples <- read_tsv("tissues.tsv")
tissues <- samples$Tissue
files <- paste0(tissues, "_quant/quant.sf")
abundance_list <- lapply(files,
	function(x) read_tsv(x, skip = 9, col_types = "cidd"))
gene_ids <- abundance_list[[1]]$`# Name`
abundance_all <- lapply(abundance_list, function(x) x$TPM) %>%
	setNames(tissues) %>%
	data.frame(ID = gene_ids, .) %>%
	left_join(ID_Name, by = 'ID') %>%
	mutate(
		Name = make_unique(Name),
		Class = factor_switch(substr(Name, 1, 3),
			"mRNA", orf = "ORF", rrn = "rRNA", trn = "tRNA")) %>%
	select(Class, ID, Name, Size, everything(), -Type)

table(Class = abundance_all$Class) %>% data_frame %>% kable
```

# Write the aggregated transcript abundance table
```{r write-abundance-table}
write.table(abundance_all, "salmon.tsv",
	quote = FALSE, sep = "\t", row.names = FALSE)
```

# Remove ORFs and all-zero rows and convert to a matrix
```{r remove-orfs}
abundance_matrix <- abundance_all %>%
	filter(Class != "ORF") %>%
	set_rownames(.$Name) %>%
	select(-Class, -ID, -Name, -Size) %>%
	.[rowSums(.) > 0,] %>%
	as.matrix
```

# Remove all-zero rows and non-coding and convert to a matrix
```{r remove-zeros}
abundance_cds_orf_matrix <- abundance_all %>%
	filter(Class %in% c("mRNA", "ORF")) %>%
	set_rownames(.$Name) %>%
	select(-Class, -ID, -Name, -Size) %>%
	.[rowSums(.) > 0,] %>%
	as.matrix
```

# Tabulate expressed genes by developmental stage
```{r tabulate-expressed-genes}
abundance_development <- abundance_all %>%
	mutate(
		Mature = Bark + FlushBud + MatureNeedle + Xylem + YoungBuds,
		Developing = Embryo + Megagametophyte + SeedGermination)

table(
	Developing = abundance_development$Developing > 0,
	Mature = abundance_development$Mature > 0,
	Class = abundance_development$Class) %>%
	addmargins
```

# Report max abundance of each tissue
```{r max-abundance}
max_abundance <- sapply(tissues,
	function(x) abundance_all[which.max(abundance_all[,x]),]) %>% t
kable(max_abundance)
```

# Heat map of transcript abundance
```{r heatmap}
NMF::aheatmap(abundance_matrix,
	fontsize = 13, cexCol = 0.7)
NMF::aheatmap(abundance_matrix,
	fontsize = 13, cexCol = 0.7,
	filename = "salmon.pdf")
```

# Heat map of transcript abundance log scale
```{r heatmap-log}
NMF::aheatmap(log10(1 + abundance_matrix),
	fontsize = 13, cexCol = 0.7)
NMF::aheatmap(log10(1 + abundance_matrix),
	fontsize = 13, cexCol = 0.7,
	filename = "salmon-log.pdf")
```

# Heat map of protein-coding transcript abundance
```{r heatmap-cds}
abundance_cds_matrix <- abundance_matrix %>%
	.[!grepl("^rrn|^trn", rownames(.)),]

NMF::aheatmap(abundance_cds_matrix,
	fontsize = 13, cexCol = 0.7)
NMF::aheatmap(abundance_cds_matrix,
	fontsize = 13, cexCol = 0.7,
	filename = "salmon-cds.pdf")
```

# Heat map of protein-coding transcript abundance log scale
```{r heatmap-cds-log}
NMF::aheatmap(log10(1 + abundance_cds_matrix),
	fontsize = 13, cexCol = 0.7)
NMF::aheatmap(log10(1 + abundance_cds_matrix),
	fontsize = 13, cexCol = 0.7,
	filename = "salmon-cds-log.pdf")
```

# Heat map of CDS and ORF transcript abundance log scale
```{r heatmap-cds-orf-log}
NMF::aheatmap(log10(1 + abundance_cds_orf_matrix),
	fontsize = 13, cexCol = 0.7)
NMF::aheatmap(log10(1 + abundance_cds_orf_matrix),
	fontsize = 13, cexCol = 0.7,
	filename = "salmon-cds-orf-log.pdf")
```

# Scatter plot of mean abundance vs length
```{r mean-abundance-vs-length}
mean_abundance = abundance_all %>%
	mutate(Mean_abundance = (
			Bark + Embryo + FlushBud + MatureNeedle +
			Megagametophyte + SeedGermination + Xylem + YoungBuds) / 8,
		Family = substr(Name, 1, 3) %>% ifelse(. == "orf", NA, .))

ggplot(mean_abundance %>% arrange(Class != "ORF")) +
	aes(x = Size, y = Mean_abundance, color = Family) +
	geom_point(alpha = 0.5) +
	geom_vline(x = 300) +
	geom_hline(y = 10) +
	scale_x_log10() +
	scale_y_log10() +
	theme_bw()
```
