# Visualize mitochondrial gene content annotated by MAKER

```{r knitr, include=FALSE}
library(knitr)
opts_chunk$set(fig.width=10, fig.height=5)
```

```{r Install libraries, eval=FALSE, include=FALSE}
install.packages('dplyr')
install.packages('lattice')
source('http://bioconductor.org/biocLite.R')
biocLite('genomeIntervals')
```

```{r Load libraries, include=FALSE}
library(dplyr)
library(genomeIntervals)
library(lattice)
```

```{r Constants}
gene.types.matrix <- t(matrix(nrow = 2, data = c(
	'atp', 'ATP synthase',
	'ccb', 'cytochrome c biogenesis',
	'ccm', 'cytochrome c maturation',
	'clp', 'maturases (clp)',
	'cob', 'complex III (ubichinol cytochrome c reductase)',
	'cox', 'complex IV (cytochrome c oxidase)',
	'mat', 'maturases',
	'mtt', 'membrane targeting and translocation',
	'nad', 'complex I (NADH dehydrogenase)',
	'ND5', 'complex I (NADH dehydrogenase 5)',
	'ori', 'origin of replication',
	'orf', 'open reading frames',
	'rpl', 'ribosomal proteins (LSU)',
	'rpo', 'RNA polymerase',
	'rps', 'ribosomal proteins (SSU)',
	'rrn', 'ribosomal RNAs',
	'sdh', 'complex II (succinate dehydrogenase)',
	'trn', 'transfer RNAs',
	'unk', 'unknown')))
gene.types <- data.frame(gene.types.matrix)
colnames(gene.types) <- c('Short.name', 'Full.name')
rownames(gene.types) <- gene.types$Short.name
```

```{r Load data}
genome.size <- 6017035
gff.all <- readGff3('pg29mt-concat.gff')
gff <- gff.all[gff.all$type %in% c('mRNA', 'tRNA'),]

data <- droplevels(data.frame(
	Type = gff$type,
	Name = getGffAttribute(gff, 'Name'),
	Size = size(gff),
	stringsAsFactors = FALSE))
data$Family = factor(substr(data$Name, 1, 3),
	levels = gene.types$Short.name,
	labels = gene.types$Full.name)
data[is.na(data$Family), 'Family'] <- 'unknown'
```

```{r Aggregate the data}
data.grouped <- droplevels(chain(data,
	group_by(Type, Family),
	summarise(
		Count = n(),
		Size = sum(Size),
		Proportion = Size / genome.size),
	filter(Size >= 1000)))

tab <- data.grouped[c('Count', 'Size', 'Proportion')]
rownames(tab) <- data.grouped$Family
tab['Sum',] <- colSums(tab) 
```

```{r Tabulate the data, results='asis'}
kable(tab)
```

```{r Stripplot of type vs size}
stripplot(Family ~ Size | Type, data,
	jitter = TRUE,
	layout = c(1, nlevels(data$Type)),
	scales = list(relation = 'free'),
	xlab = 'Size (bp)')
```

```{r Barchart type vs count}
barchart(Family ~ Count, data.grouped,
	origin = 0)

barchart(Type ~ Count, data.grouped, group = Family,
	stack = TRUE, auto.key = list(space = 'right'))

barchart(~ Count, data.grouped, group = Family,
	stack = TRUE, auto.key = list(space = 'right'))
```

```{r Barchart type vs size}
barchart(Family ~ Size / 1e3, data.grouped,
	origin = 0,
	xlab = 'Size (kbp)')

barchart(~ Size / 1e3, data.grouped, group = Family,
	stack = TRUE, auto.key = list(space = 'right'),
	xlab = 'Size (kbp)')
```
