---
title: Mitochondrial gene content annotated by MAKER
author: Shaun Jackman
---

```{r knitr, include=FALSE}
library(knitr)
opts_chunk$set(tidy = FALSE, fig.width = 10, fig.height = 5)
```

```{r Install libraries, eval=FALSE, include=FALSE}
install.packages('dplyr')
install.packages('ggplot2')
install.packages('lattice')
source('http://bioconductor.org/biocLite.R')
biocLite('genomeIntervals')
```

```{r Load libraries, include=FALSE}
library(dplyr)
library(genomeIntervals)
library(ggplot2)
library(lattice)
```

# Gene families
```{r Gene families}
gene.types.matrix <- t(matrix(nrow = 2, data = c(
	'atp', 'ATP synthase',
	'ccb', 'cytochrome c biogenesis',
	'ccm', 'cytochrome c maturation',
	'clp', 'maturases (clp)',
	'cob', 'complex III (ubichinol cytochrome c reductase)',
	'cox', 'complex IV (cytochrome c oxidase)',
	'dpo', 'DNA polymerase',
	'mat', 'maturases',
	'mtt', 'membrane targeting and translocation',
	'nad', 'complex I (NADH dehydrogenase)',
	'nd5', 'complex I (NADH dehydrogenase subunit 5)',
	'ori', 'origin of replication',
	'orf', 'open reading frames',
	'rna', 'DNA-dependent RNA polymerase',
	'rpl', 'ribosomal proteins (LSU)',
	'rpo', 'RNA polymerase',
	'rps', 'ribosomal proteins (SSU)',
	'rrn', 'ribosomal RNAs',
	'sdh', 'complex II (succinate dehydrogenase)',
	'trn', 'transfer RNAs',
	'ymf', 'uncharacterized mitochondrial protein',
	# Chloroplast
	'psa', 'photosystem I (plastidial)',
	'psb', 'photosystem II (plastidial)',
	# Other
	'bla', 'Beta-lactamase (bacterial)',
	'tns', 'transposase (transposon)',
	'unk', 'unknown')))
gene.types <- data.frame(gene.types.matrix)
colnames(gene.types) <- c('Short.name', 'Full.name')
rownames(gene.types) <- gene.types$Short.name
```

# Load data
```{r Load data}
gff.all <- readGff3('pg29mt-scaffolds.gff')
gff <- gff.all[gff.all$type %in% c('mRNA', 'rRNA', 'tRNA'),]

data <- droplevels(data.frame(
	Type = gff$type,
	Name = getGffAttribute(gff, 'Name'),
	Size = size(gff),
	stringsAsFactors = FALSE))
data$Gene <- factor(tolower(sub('trn.*', 'trn',
	sub('[-_|].*', '', data$Name))))
data$Family.short <- tolower(substr(data$Name, 1, 3))
data$Family = factor(data$Family.short,
	levels = gene.types$Short.name,
	labels = gene.types$Full.name)
data[is.na(data$Family), 'Family'] <- 'unknown'
```

# Aggregate the data
```{r aggregate}
data.grouped.with.orf <- data %>%
	group_by(Type, Family.short, Family) %>%
	summarise(
		Count = n(),
		Size = sum(Size)) %>%
	droplevels

data.grouped <- data.grouped.with.orf %>% filter(Family.short != "orf")
```

# Tabulate the data
```{r tabulate, results='asis'}
tab <- data.grouped.with.orf[c('Count', 'Size')]
rownames(tab) <- data.grouped.with.orf$Family
tab['Sum',] <- colSums(tab)
kable(tab)
```

# Histogram of size (â‰¤ 1000 bp)
```{r gene-size-density}
ggplot(data = data) +
	aes(x = Size) +
	geom_histogram(binwidth = 25) +
	xlim(0, 1000) +
	xlab("Size (bp)")
```

# Stripplot of type vs size
```{r Stripplot of type vs size}
stripplot(Family ~ Size, data,
	jitter = TRUE,
	xlab = 'Size (bp)')
```

# Density plot of type vs size
```{r density-type-vs-size}
ggplot(data = data) +
	aes(x = Size) +
	facet_wrap(~ Family.short) +
	geom_density() +
	xlab("Size (bp)")
```

# Barchart of type vs count
```{r Barchart of type vs count}
barchart(Family ~ Count, data.grouped,
	origin = 0)

barchart(Type ~ Count, data.grouped, group = Family,
	stack = TRUE, auto.key = list(space = 'right'))
```

# Barchart of type vs size
```{r Barchart type vs size, fig.height = 8}
barchart(Family ~ Size / 1e3,
	data %>% filter(Family.short != "orf") %>% droplevels,
	group = Gene,
	stack = TRUE, auto.key = list(space = 'right'),
	xlab = 'Size (kbp)')

barchart(~ Size / 1e3, data.grouped, group = Family,
	stack = TRUE, auto.key = list(space = 'right'),
	xlab = 'Size (kbp)')
```

# Write the tabulated data to a file
```{r Write the tabulated data to a file}
write.table(data.grouped.with.orf, 'pg29mt-scaffolds.gff.tsv',
	quote = FALSE, sep = '\t', row.names = FALSE)
```
