---
title: Mitochondrial gene content annotated by MAKER
author: Shaun Jackman
output:
  html_document:
    keep_md: yes
    toc: yes
---

```{r knitr, include=FALSE}
library(knitr)
opts_chunk$set(tidy = FALSE, fig.width = 10, fig.height = 5)
```

```{r Install libraries, eval=FALSE, include=FALSE}
install.packages('dplyr')
install.packages('ggplot2')
install.packages('lattice')
source('http://bioconductor.org/biocLite.R')
biocLite('genomeIntervals')
```

```{r Load libraries, include=FALSE}
library(dplyr)
library(genomeIntervals)
library(ggplot2)
library(lattice)
```

# Functions
```{r functions}
# Set the row names of x to value
set_row_names <- function(x, value) { row.names(x) <- value; x }
```

# Gene families
```{r Gene families}
gene.types.matrix <- t(matrix(nrow = 2, data = c(
	'atp', 'ATP synthase',
	'ccb', 'cytochrome c biogenesis',
	'ccm', 'cytochrome c maturation',
	'clp', 'maturases (clp)',
	'cob', 'complex III (ubichinol cytochrome c reductase)',
	'cox', 'complex IV (cytochrome c oxidase)',
	'dpo', 'DNA polymerase',
	'mat', 'maturases',
	'mtt', 'membrane targeting and translocation',
	'nad', 'complex I (NADH dehydrogenase)',
	'ori', 'origin of replication',
	'orf', 'open reading frames',
	'psa', 'photosystem I',
	'rna', 'DNA-dependent RNA polymerase',
	'rpl', 'ribosomal proteins (LSU)',
	'rpo', 'RNA polymerase',
	'rps', 'ribosomal proteins (SSU)',
	'sdh', 'complex II (succinate dehydrogenase)',
	'ymf', 'uncharacterized mitochondrial protein',
	'oth', 'other genes',
	'rrn', 'ribosomal RNAs',
	'trn', 'transfer RNAs')))
gene.types <- data.frame(gene.types.matrix)
colnames(gene.types) <- c('Family.short', 'Family')
rownames(gene.types) <- gene.types$Short.name
```

# Load data
```{r Load data}
gff.all <- readGff3("pg29mt-scaffolds.gff", isRightOpen = FALSE)
gff <- gff.all[gff.all$type %in% c('mRNA', 'rRNA', 'tRNA'),]

data <- droplevels(data.frame(
	Type = gff$type,
	ID = getGffAttribute(gff, 'ID'),
	Name = getGffAttribute(gff, 'Name'),
	Size = size(gff),
	stringsAsFactors = FALSE))
data$Gene <- factor(
	sub('RNA', 'RNA_pol',
	sub('trn.*', 'trn*',
	sub('ymf.*', 'ymf*',
	sub('[-_|].*', '', data$Name)))))
data$Family.short <- tolower(substr(data$Name, 1, 3))
data$Family = factor(data$Family.short,
	levels = gene.types$Family.short,
	labels = gene.types$Family)
data[is.na(data$Family), 'Family'] <- 'other genes'
```

# Aggregate the data
```{r aggregate}
data_grouped <- data %>%
	group_by(Type, Family) %>%
	summarise(
		Count = n(),
		Size = sum(Size)) %>%
	droplevels %>%
	right_join(gene.types, .)
data_grouped_without_orf <- data_grouped %>% filter(Family.short != "orf")
```

# Tabulate genes by type
```{r tabulate-genes-by-type}
tab <- data %>%
	group_by(Type) %>%
	summarise(
		Count = n(),
		Size = sum(Size)) %>%
	set_row_names(.$Type) %>%
	select(Count, Size)
tab['Sum',] <- colSums(tab)
kable(tab)
```

# Tabulate genes by family
```{r tabulate-genes-by-family}
tab <- data_grouped %>%
	set_row_names(.$Family) %>%
	ungroup %>%
	select(Count, Size)
tab['Sum',] <- colSums(tab)
kable(tab)
```

# Tabulate known genes without ORFs and other genes
```{r tabulate-known-genes}
tab <- data_grouped %>%
	filter(!Family.short %in% c("orf", "oth")) %>%
	set_row_names(paste(.$Family.short, .$Family)) %>%
	ungroup %>%
	select(Count, Size)
tab['Sum',] <- colSums(tab)
kable(tab)
```

# Tabulate protein-coding genes (mRNA) by family
```{r tabulate-mrna-by-family}
tab <- data_grouped %>%
	filter(!Family.short %in% c("orf", "oth", "rrn", "trn")) %>%
	set_row_names(paste(.$Family.short, .$Family)) %>%
	ungroup %>%
	select(Count, Size)
tab['Sum',] <- colSums(tab)
kable(tab)
```

# Tabulate protein-coding genes (mRNA) by gene
```{r tabulate-mrna-by-gene}
ignore_families <- c("orf", "bla", "tns", "rrn", "trn")

data %>%
	filter(!Family.short %in% ignore_families) %>%
	summarize(Count = n(), Distinct = n_distinct(Gene), Size = sum(Size)) %>%
	kable

data %>%
	filter(!Family.short %in% ignore_families) %>%
	group_by(Gene) %>%
	summarize(Count = n(), Size = sum(Size)) %>%
	kable
```

# Tabulate open reading frames (ORF)
```{r tabulate-orf}
data_grouped %>% filter(Family.short == "orf") %>% kable
```

# Tabulate open reading frames (ORF) at least 300 bp
```{r tabulate-orf-300bp}
data %>%
	filter(Family.short == "orf" & Size >= 300) %>%
	group_by(Type, Family) %>%
	summarise(
		Count = n(),
		Size = sum(Size)) %>%
	right_join(gene.types, .) %>%
	kable
```

# List other genes
```{r list-other-genes}
data %>% filter(Family == "other genes") %>% kable
```

# Tabulate intons
```{r tabulate-introns}
introns_gff <- gff.all[gff.all$type == "intron",]
introns <- data.frame(
		getGffAttribute(introns_gff, 'Parent'),
		Size = size(introns_gff),
		stringsAsFactors = FALSE) %>%
	left_join(data, by = c("Parent" = "ID")) %>%
	rename(Intron_size = Size.x, Gene_size = Size.y) %>%
	select(Type, Name, Intron_size, Gene_size) %>%
	arrange(Type, Name) %>%
	mutate(Name = sub('[|].*', '', Name))

introns %>%
	group_by(Type) %>%
	summarize(Count = n(), Distinct = n_distinct(Name), Size = sum(Intron_size))

introns %>%
	group_by(Type, Name) %>%
	summarize(Count = n(), Size = sum(Intron_size)) %>%
	kable
```

# Histogram of size (â‰¤ 1000 bp)
```{r gene-size-density}
ggplot(data = data) +
	aes(x = Size) +
	geom_histogram(binwidth = 50) +
	scale_x_continuous(limits = c(0, 1000), breaks = seq(0, 1000, 100)) +
	xlab("Size (bp)")
```

# Stripplot of type vs size
```{r Stripplot of type vs size}
stripplot(Family ~ Size, data,
	jitter = TRUE,
	xlab = 'Size (bp)')
```

# Density plot of type vs size
```{r density-type-vs-size}
ggplot(data = data) +
	aes(x = Size) +
	facet_wrap(~ Family.short) +
	geom_density() +
	xlab("Size (bp)")
```

# Barchart of type vs count
```{r Barchart of type vs count}
barchart(Family ~ Count, data_grouped_without_orf,
	origin = 0)

barchart(Type ~ Count, data_grouped_without_orf, group = Family,
	stack = TRUE, auto.key = list(space = 'right'))
```

# Barchart of type vs size
```{r barchart-type-vs-size, fig.height = 9}
ignore_families <- c("orf", "bla", "tns")
barchart(Family ~ Size / 1e3,
	data %>% filter(!Family.short %in% ignore_families) %>% droplevels,
	group = Gene,
	stack = TRUE, auto.key = list(reverse.rows = TRUE, space = 'right'),
	xlab = 'Size (kbp)')

barchart(~ Size / 1e3, data_grouped_without_orf, group = Family,
	stack = TRUE, auto.key = list(space = 'right'),
	xlab = 'Size (kbp)')
```

# Write the tabulated data to a file
```{r Write the tabulated data to a file}
write.table(data_grouped, 'pg29mt-scaffolds.gff.tsv',
	quote = FALSE, sep = '\t', row.names = FALSE)
```
